### PR TITLE
Issue #4 updated codes to make python3 compatibility.

### PR DESCRIPTION
**Pull request by nazrulworld**
* flake8 cleanup [nazrulworld]

### CODE REVIEW

**Comment by jamadden** [comment]
Thanks for the PR! It will need a few more things to make it complete:

- Please add the versions of Python 3 to .travis.yml and make sure the tests pass there ( [zope.schema is a good example](https://github.com/zopefoundation/zope.schema/blob/master/.travis.yml) )
- Likewise for tox.ini
- The classifiers in setup.py need to be updated as well ([example](https://github.com/zopefoundation/zope.schema/blob/master/setup.py#L63))

**Comment by nazrulworld** [comment]
Thanks for your valuable feedback! I will do accordingly

**Comment by nazrulworld** [comment]
@jamadden added python2/3 compatibility codes added.

File: src/z3c/saconfig/zcml.py [hunk 1]
```diff
@@ -1,111 +1,128 @@
...
-import utility
-import interfaces
 
 class IEngineDirective(zope.interface.Interface):
     """Registers a database engine factory."""
 
     url = zope.schema.URI(
-        title=u"Database URL",
-        description=u"e.g. 'sqlite:///:memory:'.",
+        title=six.text_type("Database URL"),
```
Comment (ale-rt) [inline src/z3c/saconfig/zcml.py]: I think this is not necessary: u"Database URL" is equal to six.text_type("Database URL") in both Python versions

File: buildout.cfg [hunk 2]
```diff
@@ -6,7 +6,8 @@ develop = .
 recipe = zc.recipe.testrunner
 eggs =
     z3c.saconfig [test]
-    pysqlite
+#    pysqlite
```
Comment (ale-rt) [inline buildout.cfg]: I see you are commenting `#    pysqlite` here but then a version pin is added below. Looks strange :)

File: src/z3c/saconfig/README.rst [hunk 3]
```diff
@@ -333,7 +334,7 @@ Configuration using ZCML
 A configuration directive is provided to register a database engine
 factory using ZCML.
 
-  >>> from cStringIO import StringIO
+  >>> from six import StringIO
```
Comment (ale-rt) [inline src/z3c/saconfig/README.rst]: Probably you should use from io import BytesIO and be sure that you are passing bytes to it.
See this patch as an example https://github.com/plone/plone.supermodel/commit/1d02b25d02b825cd64fa95760b0abee1b5055e7e#diff-bc1d78c2a92695a5b3d8a63f5ea6ab0fR16

**Review by ale-rt — CHANGES_REQUESTED** [review]
Thanks a lot :)
It looks really well done to me, there are just a few things to tune up.

In addition be sure you have signed the Zope contributor agreemeent: https://docs.zope.org/developer/becoming-a-committer.html

File: src/z3c/saconfig/zcml.py (hunk 1 above)
Comment (nazrulworld) [reply src/z3c/saconfig/zcml.py]: You are right in Python 3.3+ it's legal again to use u''. I will revert that

**Comment by nazrulworld** [comment]
@ale-rt I have mailed with Signed copy Zope Committer Agreement

File: .travis.yml [hunk 4]
```diff
@@ -2,11 +2,29 @@ language: python
 sudo: false
 python:
     - 2.7
+    - 3.5
+    - 3.6
```
Comment (mgedmin) [inline .travis.yml:6]: You can add 3.7 if you do the magic dance of

```
matrix:
    include:
        - python: 3.7
          dist: xenial
          sudo: required
```

Comment (mgedmin) [reply .travis.yml:6]: (It's fine if you don't want to do that at this time.)

File: CHANGES.rst [hunk 5]
```diff
@@ -4,7 +4,8 @@ z3c.saconfig
 0.15 (unreleased)
 =================
 
-- Nothing changed yet.
+- all tests are ported to python3 compatibility [nazrulworld]
```
Comment (mgedmin) [inline CHANGES.rst]: Would be nice to mention which specific Python 3 versions are supported (AFAICT they're 3.5 and 3.6).

Usually we phrase this as

> - Added Python 3.5 and 3.6 compatibility

File: setup.py [hunk 6]
```diff
@@ -1,39 +1,47 @@
...
+                   'Programming Language :: Python :: 3.6',
                    'Topic :: Internet :: WWW/HTTP',
                    'Topic :: Software Development :: Libraries',
                    'Topic :: Database',
                    ],
       keywords='zope relational sqlalchemy component integration',
       author='Martijn Faassen',
       author_email='faassen@startifact.com',
-      url='http://pypi.python.org/pypi/z3c.saconfig',
+      url='https://pypi.org/project/z3c.saconfig/',
```
Comment (mgedmin) [inline setup.py]: IMHO a GitHub link would be better -- this is the URL shown on PyPI, and it's a little pointless to have a PyPI page point back to itself.

File: src/z3c/saconfig/README.rst [hunk 7]
```diff
@@ -333,41 +334,41 @@ Configuration using ZCML
...
-  >>> from cStringIO import StringIO
+  >>> from io import BytesIO
   >>> from zope.configuration import xmlconfig
   >>> import z3c.saconfig
   >>> xmlconfig.XMLConfig('meta.zcml', z3c.saconfig)()
 
 Let's try registering the directory again.
 
-  >>> xmlconfig.xmlconfig(StringIO("""
+  >>> xmlconfig.xmlconfig(BytesIO(six.b("""
```
Comment (mgedmin) [inline src/z3c/saconfig/README.rst]: No need for `six.b`, `b"..."` literals work across all currently supported Python versions.

File: src/z3c/saconfig/README.rst [hunk 8]
```diff
@@ -141,12 +141,13 @@ When a new engine is created by an ``EngineFactory``, an
...
+  >>> import six
   >>> from z3c.saconfig.interfaces import IEngineCreatedEvent
   >>> @component.adapter(IEngineCreatedEvent)
   ... def createdHandler(event):
-  ...     print "created engine"
-  ...     print "args:", event.engine_args
-  ...     print "kw:", event.engine_kw
+  ...     six.print_("created engine")
+  ...     six.print_("args:", event.engine_args)
+  ...     six.print_("kw:", event.engine_kw)
```
Comment (mgedmin) [inline src/z3c/saconfig/README.rst]: There's a way of using `from __future__ import print_function` in doctests, then you could use `print()`.

(Or if you refrain from using multiple args, you can always do `print("args: {}".format(event.engine_args))` etc. and it'll work on both major Python versions without future imports.)

File: src/z3c/saconfig/tests.py [hunk 9]
```diff
@@ -7,36 +7,37 @@
...
 from zope.testing import cleanup
 from zope.testing.cleanup import addCleanUp
 
-from zope import component
-from zope.component import registry
+import doctest
+import os
+import six
+import unittest
 import zope.component.eventtesting
```
Comment (mgedmin) [inline src/z3c/saconfig/tests.py:22]: This seems wrong: PEP-8 suggests putting standard library imports (doctest, os, unittest) in a group above 3rd-party library imports (zope.*).

File: src/z3c/saconfig/utility.py [hunk 10]
```diff
@@ -1,19 +1,24 @@
...
+from zope.event import notify
+from zope.interface import implementer
+from zope.sqlalchemy import ZopeTransactionExtension
 
-import time
-import thread
-import threading
 import sqlalchemy
+import threading
+import time
```
Comment (mgedmin) [inline src/z3c/saconfig/utility.py:15]: (Likewise)

File: src/z3c/saconfig/zcml.py [hunk 11]
```diff
@@ -1,76 +1,85 @@
...
 
     convert_unicode = zope.schema.Bool(
         title=u'Convert all string columns to unicode',
-        description=u'This setting makes the SQLAlchemy String column type '
-                    u'equivalent to UnicodeString. Do not use this unless '
-                    u'there is a good reason not to use standard '
+        description=u'This setting makes the SQLAlchemy String column type ' +
+                    u'equivalent to UnicodeString. Do not use this unless ' +
+                    u'there is a good reason not to use standard ' +
                     u'UnicodeString columns',
```
Comment (mgedmin) [inline src/z3c/saconfig/zcml.py:37]: Why the `+`s?  String concatenation is a perfectly cromulent syntactic feature.

File: src/z3c/saconfig/zcml.py [hunk 12]
```diff
@@ -116,25 +128,30 @@ def engine(_context, url, name=u"", convert_unicode=False, echo=None, setup=None
...
         component=factory,
         permission=PublicPermission,
         name=name)
-    
+
     if setup:
+        try:
+            setup = setup.decode()
+        except AttributeError:
+            pass
```
Comment (mgedmin) [inline src/z3c/saconfig/zcml.py]: I don't like this.

It would be better to

```
if isinstance(setup, bytes):
    setup = setup.decode()
```

**Review by mgedmin — APPROVED** [review]
LGTM, with minor possible improvements.

File: src/z3c/saconfig/README.rst (hunk 8 above)
Comment (mgedmin) [reply src/z3c/saconfig/README.rst]: (There are two ways of using print_function in doctests: you can just write it out in the doctest itself like [zc.relation does](https://github.com/zopefoundation/zc.relation/blob/master/src/zc/relation/README.rst), or you can stuff it into globs like [zope.mimetype does](https://github.com/zopefoundation/zope.mimetype/blob/master/src/zope/mimetype/tests/test_doctests.py#L60).)

File: src/z3c/saconfig/tests.py [hunk 13]
```diff
@@ -48,53 +49,67 @@ def tearDownReadMe(test):
...
 
 # make sure hooks get cleaned up after tests are run
+
+
 addCleanUp(resetHooks)
 
+
 def engine_subscriber(engine):
-    print "got: %s " % engine
+    six.print_('got: {0!s} '.format(engine))
```
Comment (mgedmin) [inline src/z3c/saconfig/tests.py]: This is the only remaining place where you use `six`, and it can be rewritten to use `print()` directly.

Comment (nazrulworld) [reply src/z3c/saconfig/tests.py]: thanks a lot :) now six dependency is wiped

**Review by mgedmin — APPROVED** [review]
Feel free to merge when your commit bit comes through!

**Review by ale-rt — APPROVED** [review]

**Comment by nazrulworld** [comment]
@ale-rt I am now approved as member as committer. Can you please initiate to merge this PR. Thanks

**Comment by jamadden** [comment]
The general policy in zopefoundation repositories is that once the PR is approved, the person that opened the PR clicks the "merge" button. This serves to prove that they've signed the committer agreement and have the correct permissions. (Put another way, we usually merge our own PRs after they've been approved by others.) Feel free to click the merge button!

**Pull request merged**